### PR TITLE
🐛 Fix CoreDNS upgrade from v1.20 to v1.21

### PR DESCRIPTION
**Pull request by fabriziopandini**
**What this PR does / why we need it**:
This PR is a backport of https://github.com/kubernetes-sigs/cluster-api/pull/4476

The upstream image for CoreDNS was renamed from 'k8s.gcr.io/coredns' to `k8s.gcr.io/coredns/coredns`, and this is impacting upgrades of workload clusters from v1.20 to v1.21. 
This PR makes KCP upgrades to account for this change.

Please note that:
- the change applies only to users relying on upstream image repository (`k8s.gcr.io`); users relying on custom repository could continue to use `myrepo.io/coredns` as image name (no changes required).
- starting from CoreDNS 1.8.0, the upstream images requires the `v` prefix on the ImageTag, but this should be taken care by the user when specifying the desired version in KCP.

**Which issue(s) this PR fixes**:
Rif https://github.com/kubernetes-sigs/cluster-api/issues/4463

/assign @vincepri 
@sbueringer

### CODE REVIEW

**Comment by vincepri** [comment]
/retitle 🐛 Fix CoreDNS upgrade from v1.20 to v1.21

File: controlplane/kubeadm/internal/workload_cluster_coredns.go [hunk 1]
```diff
@@ -19,6 +19,9 @@ package internal
 import (
 	"context"
 	"fmt"
+	"strings"
+
+	"github.com/blang/semver"
```
Comment (vincepri) [inline controlplane/kubeadm/internal/workload_cluster_coredns.go:25]:
```suggestion
```

**Review by vincepri — APPROVED** [review]
/approve

File: controlplane/kubeadm/internal/workload_cluster_coredns_test.go [hunk 2]
```diff
@@ -280,6 +286,131 @@ kind: ClusterConfiguration
...
+			name: "updates everything successfully to v1.8.0 with a custom repo should not change the image name",
+			kcp: &controlplanev1.KubeadmControlPlane{
+				Spec: controlplanev1.KubeadmControlPlaneSpec{
+					KubeadmConfigSpec: cabpkv1.KubeadmConfigSpec{
+						ClusterConfiguration: &kubeadmv1.ClusterConfiguration{
+							DNS: kubeadmv1.DNS{
+								Type: kubeadmv1.CoreDNS,
+								ImageMeta: kubeadmv1.ImageMeta{
+									// provide an newer image to update to
+									ImageRepository: "k8s.gcr.io/some-repo",
```
Comment (andrewsykim) [inline controlplane/kubeadm/internal/workload_cluster_coredns_test.go:301]: Wondering if this test should probably use an image repo that is not k8s.gcr.io in this test

Comment (sbueringer) [reply controlplane/kubeadm/internal/workload_cluster_coredns_test.go:301]: Would have been a bit better

**Pull request merged**